abi.html

```diff
@@ -5827,6 +5957,298 @@ <h5><a href="#mangling.declaration-reference">5.1.6.2 References to declared ent
+</code></font></pre>
+
+<p>
+Next, if the constant is a pointer, the reference expression is
+wrapped in a unary <code>&</code> operator (that is, prefixed with
+<code class=mangle>ad</code>.  The notional type of this expression
+is a pointer to the notional type of the previous reference expression.
+
+<p>
+Finally, if precise typing is required and the notional type of the
+current reference expression doesn't match the type of the constant,
+the reference expression is wrapped in a C-style cast to the constant
+type (that is, prefixed with <code class=mangle>cv &lt;type&gt;</code>).
+
+<a name="mangling.member-pointer-reference">
+<h5><a href="#mangling.member-pointer-reference">5.1.6.4 Member pointer reference expressions</a></h5>
+</a>
+
+<p>
+Constant evaluation may result in a member pointer to a specific
+member, possibly converted to a different type.  As with other constant
+manglings, this value is identified in the mangling as if it were written
+with a certain expression, called the reference expression.  The exact expression depends on whether <a href="#mangling.constant-values">precise
+typing</a> is required.
+
+<p>
+The initial reference expression applies the prefix operator
+<code>&</code> to the member declaration as if it were a regular
+function or variable.  The notional type of this expression is a
+member pointer type using the declaring class and the declared type of
+the member. For example:
+
+<code><pre>
+struct A {
+    void foo();
+};
+template &lt;void (A::*)()&gt; void f();
+
+f&lt;&amp;A::foo&gt; // _Z1fIXadL_ZN1A3fooEvEEEvv
+</pre></code>
+
+<p>
+If the class type of the constant is not the declaring class of the
+member, the reference expression is wrapped in a special expression
+used only for this purpose, which encodes the <code>this</code>-adjustment
+that must be applied to a pointer to the class type of the constant
+in order to produce a pointer to the declaring class of the member.
+The offset is omitted if zero.  The notional type of this expression
+is a member pointer type using the class type of the constant and the
+declared type of the member.
+
+<pre><font color=blue><code>
+  &lt;expression&gt; ::= mc &lt;<i>class</i> <a href="#mangle.type">type</a>&gt; &lt;<a href="#mangle.expression">expression</a>&gt; [&lt;<i>offset</i> <a href="#mangle.number">number</a>&gt;] E
+</code></font></pre>
+
+<p>
+Finally, if precise typing is required and the notional type of the
+current reference expression is not the type of the constant, the
+reference expression is wrapped in a C-style cast to the constant
+type (that is, prefixed with <code class=mangle>cv &lt;type&gt;</code>).
+This can happen because <code>static_cast</code> may add qualifiers
+to the member type.
+
+<code><pre>
+struct A {
+  char x;
+  void foo();
+};
+struct B {
+  char y;
+};
+
+struct C : A, B {};
+template &lt;void (C::*)()&gt; void f();
+
+f&lt;&amp;A::foo&gt; // _Z1fIXmcM1CFvvEadL_ZN1A3fooEvE1EEEvv
```

Comment:
Should `mc` mangling contain class type or pointer-to-member type, as in this example?